### PR TITLE
fix mounting Hostpath devices

### LINKED ISSUE #79090: k8s api doesn't support the --device args which support by the docker api。
**I want mount a char device into container，but not grant privilege to the container**:
1. I use hostpath binding a char device to container。
2. if i doesn't grant privilege to the container， i can't open the char device in the docker container。
3.  if i grant privilege to the container， char device can open in the docker container。

**Docker command client support --device，but i can't find this api in k8s。**

### PR DESCRIPTION
**Pull request by dixudx**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
HostPathVolume does support binding a char/block device to container. Without privilege mode, the device inside the container can not be opened/wrote.

**Which issue(s) this PR fixes**:

Fixes #79090

**Special notes for your reviewer**:
/cc msau42 rootfs 
/cc kubernetes/sig-node-pr-reviews 

**Does this PR introduce a user-facing change?**:

```release-note
fix mounting Hostpath devices by lifting HostPathCharDev/HostPathBlockDev to volume devices
```

### CODE REVIEW

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 1]
```diff
@@ -757,6 +787,8 @@ func (m *kubeGenericRuntimeManager) SyncPod(pod *v1.Pod, podStatus *kubecontaine
 	// Step 6: start containers in podContainerChanges.ContainersToStart.
 	for _, idx := range podContainerChanges.ContainersToStart {
 		container := &pod.Spec.Containers[idx]
+		liftHostPathDeviceMountToVolumeDevices(pod, container)
```
Comment (vikaschoudhary16) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: Can you please add a unit test for this function too?

**Comment by msau42** [comment]
Is the device plugin work supposed to address this use case? cc @RenaudWasTaken @vishh

**Comment by warlice** [comment]
please merge quickly！ thanks。

**Comment by wormwang** [comment]
mounting Hostpath devices is a basic feature for docker & k8s. pls merge it soon.

**Comment by dixudx** [comment]
kindly ping @msau42 @yujuhong for review.

This is a bug for hostpath devices. We should merge it ASAP.

**Comment by msau42** [comment]
I don't think we should add this special case to hostpath volumes to support accessing arbitrary devices, which is a privileged operation. We now have 2 supported ways in Kubernetes to make devices available to containers:
- CSI drivers
- Device plugin

**Comment by dixudx** [comment]
> We now have 2 supported ways in Kubernetes to make devices available to containers

@msau42  Yeah, I know. But for hostpath, it is a bug. Many people are still using hostpath to mount volumes.

**Comment by wormwang** [comment]
We are still using hostpath to mount user extended  volumes .pls support the feature

**Comment by dixudx** [comment]
> We are still using hostpath to mount user extended volumes .pls support the feature

@wormwang This is exactly what this PR is trying to address.

@msau42 PTAL. Thanks.

**Comment by msau42** [comment]
@kubernetes/sig-storage-pr-reviews 
@kubernetes/sig-node-pr-reviews 
cc @tallclair 

Allowing containers to specify arbitrary devices is very dangerous and I don't think we should be making it easier for unprivileged containers to do that. Like I said earlier, we have a number of other supported ways to give devices to containers that gives admins control over what gets exposed:
* CSI drivers
* device plugins
* Local PV

I would prefer using those existing methods instead of adding a 4th.

**Comment by wormwang** [comment]
--device is a feature of docker, so user can control and use it seamlessly. Other 3 methods is depend on extend of k8s, are more complex. and, Local PV can not support customized block devices.

**Comment by warlice** [comment]
> 
> 
> **What type of PR is this?**
> /kind bug
> 
> **What this PR does / why we need it**:
> HostPathVolume does support binding a char/block device to container. Without privilege mode, the device inside the container can not be opened/wrote.
> 
> **Which issue(s) this PR fixes**:
> 
> Fixes #79090
> 
> **Special notes for your reviewer**:
> /cc msau42 rootfs
> /cc kubernetes/sig-node-pr-reviews
> 
> **Does this PR introduce a user-facing change?**:
> 
> ```
> fix mounting Hostpath devices by lifting HostPathCharDev/HostPathBlockDev to volume devices
> ```

hello，we use the fix code into v16.3，but it doesn't work。 here is the log：
 kubectl describe pod
cannot find volume "devmychar" to pass into container "0f2cedd3-a2d4-4838-82c1-082ec9621ade-35494868-1。

here is the log of kubelet 
Dec 19 12:02:08 k8snode kubelet[472695]: E1219 12:01:57.817825  472695 pod_workers.go:191] Error syncing pod 3b1afd80-74c4-4a43-8f93-863618da157e ("mytestdefaut(3b1afd80-74c4-4a43-8f93-863618da157e)"), skipping: failed to "StartContainer" for "0f2cedd3-a2d4-4838-82c1-082ec9621ade-35494868-1" with CreateContainerConfigError: "cannot find volume \"devmychar\" to pass into container \"0f2cedd3-a2d4-4838-82c1-082ec9621ade-35494868-1\""

Dec 19 12:02:08 k8snode kubelet[472695]: E1219 12:02:08.813859  472695 kubelet_pods.go:108] Block volume cannot be satisfied for container "0f2cedd3-a2d4-4838-82c1-082ec9621ade-35494868-1", because the volume is missing or the volume mapper is nil: {Name:devmychar DevicePath:/dev/mychar}

**Comment by dixudx** [comment]
@warlice Please have another try. Thanks.

**Comment by dixudx** [comment]
kindly ping @msau42 @yujuhong  for a review

**Comment by mderoy** [comment]
I think closing this ticket is a mistake. 

Search google for "kubernetes character device operation not permitted" and you'll find loads of people trying to access devices in unprivileged pods via hostpath and getting unexpected behavior (operation not permitted errors, but device has all appropriate permissions, inodes, etc). 

Local PV and CSI both involve creating and managing PVCs, which is a pain. The device plugin approach works, but is not feasible for all cases (our application has a kernel module which creates devices after binding them to some disks so the devices are not available until this binding operation takes place, which for us is done by a privileged initcontainer....you get a chicken and egg issue where you can't schedule the pod because devicemanager reports devices don't exist but the initcontainer that sets up the devices can't run and set those devices up.

**Comment by mderoy** [comment]
@msau42 CSI drivers don't appear to support mounting character devices into pods. MapBlockVolume will fail at AttachFileDevice because the device is not a loopback device. This is more reason to consider this option as there is no appropriate workarounds, and the expected behavior is for these devices to work. Perhaps this should be retooled with appropriate scc permissions guarding it?

**Comment by msau42** [comment]
@mderoy thank you for trying the CSI approach and finding the issue. I did not realize that we had made some assumptions that the devices had to be block devices, but looking back that makes sense for storage. I think for character devices we should work with sig-node to find the most appropriate mechanism to support this. I still remain concerned from a security standpoint on allowing unprivileged users to access any device on the host. Is this something you would be willing to discuss in a future sig-node meeting?

**Comment by mderoy** [comment]
@msau42 I think the concerns you have for unprivileged access to host devices is quite valid, but I also think that it could be guarded by a special SCC privilege...like 'allowHostDevices' (We use openshift k8s so they already have some additional scc stuff around hostpath already) which when enabled gives the expected device access. (of course this probably opens up the flood gates when every hardware manufacturer starts telling people that to use their GPUs they need to enable this "unsafe" privilege....I can see why you're hesitant)

I'd be happy to bring this up to sig-node...I suppose the place to start is their slack?

**Pull request closed without merging**